File: src/main/scala/com/resonant/core/graph/internal/GraphConnect.java
package com.resonant.core.graph.internal;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * A graph that contains nodes, each with its ability to connect to other nodes.
 * @author Calclavia
 */
public abstract class GraphConnect<N extends Node> implements Graph<N> {

	//A list of nodes in the graph
	protected final List<N> nodes = new ArrayList<>();

	protected boolean[][] adjMat = null;

	@Override
	public void add(N node) {
		if (!nodes.contains(node)) {
			nodes.add(node);
		}
	}

	@Override
	public void remove(N node) {
		nodes.remove(node);
	}

	public int id(Object node) {
		assert nodes.contains(node);
		return nodes.indexOf(node);
	}

	@Override
	public Collection<N> nodes() {
		return nodes;
	}

	public boolean isConnected(N a, N b) {
		return adjMat[id(a)][id(b)];
	}

	public void connect(N a, N b) {
		adjMat[id(a)][id(b)] = true;
	}

	@Override
	public void build() {
		/**
		 * Builds the adjacency matrix
		 */
		adjMat = new boolean[nodes.size()][nodes.size()];

		for (N node : nodes) {
			for (Object con : node.connections()) {
				if (nodes.contains(con)) {
					connect(node, (N) con);
				}
			}
		}
	}
}

File: src/main/scala/com/resonant/core/graph/internal/GraphConnect.scala
package com.resonant.core.graph.internal

import java.util.{List => JLIst}

import com.resonant.lib.math.matrix.AdjacencyMatrix

import scala.collection.JavaConversions._
/**
 * A graph that contains getNodes, each with its ability to connect to other getNodes.
 * @author Calclavia
 */
abstract class GraphConnect[N <: Node[_]] extends Graph[N] {
	protected var nodes = Set.empty[N]
	protected[graph] var adjMat: AdjacencyMatrix[N] = null

	def add(node: N) {
		nodes += node
	}

	def remove(node: N) {
		nodes -= node
	}

	//TODO: Collection?
	override def getNodes: JLIst[N] = nodes.toList

	def build() {
		adjMat = new AdjacencyMatrix[N](nodes, nodes)

		for (node <- nodes) {
			for (con <- node.connections) {
				if (nodes.contains(con)) {
					adjMat(node, con.asInstanceOf[N]) = true
				}
			}
		}
	}
}